MelonLoader.Installer/InstallerUtils.cs

```diff
@@ -33,35 +54,77 @@ static InstallerUtils()
 
         if (length > 0)
         {
             destination.SetLength(length);
+            var position = 0;
+            var buffer = new byte[1024 * 16];
+            while (position < destination.Length - 1)
+            {
+                var read = await content.ReadAsync(buffer, 0, buffer.Length);
+                await destination.WriteAsync(buffer, 0, read);
+
+                position += read;
+
+                onProgress?.Invoke(position / (double)(destination.Length - 1), null);
+            }
         }
         else
         {
             await content.CopyToAsync(destination);
-            return null;
         }
-
-        var position = 0;
-        var buffer = new byte[1024 * 16];
-        while (position < destination.Length - 1)
+        
+        // Save cache
+        if (useCache)
         {
-            var read = await content.ReadAsync(buffer, 0, buffer.Length);
-            await destination.WriteAsync(buffer, 0, read);
-
-            position += read;
-
-            onProgress?.Invoke(position / (double)(destination.Length - 1), null);
+            try
+            {
+                Directory.CreateDirectory(Path.Combine(Config.CacheDir, "Cache", parentDirectory));
+                await using var fsIn = File.OpenWrite(fileCache);
+                destination.Seek(0, SeekOrigin.Begin);
+                await destination.CopyToAsync(fsIn);
+            }
+            catch
+            {
+                // Failed to save cache
+            }
         }
-
         return null;
     }
 
+    public static async Task<string?> DownloadFileAsync(string url, Stream destination, bool useCache, InstallProgressEventHandler? onProgress)
+    {
+        // Get archive
+        var result = await FetchFile(url, destination, useCache, onProgress);
+        if (result != null)
+            return $"Failed to fetch file from {url}: {result}";
+
+        destination.Seek(0, SeekOrigin.Begin);
+        
+        // Get checksum
+        var checksumUrl = url.Replace(".zip", ".sha512");
+        using var checksumStr = new MemoryStream();
+        result = await FetchFile(checksumUrl, checksumStr, useCache, onProgress);
+
+        // Checksum fetch failed, skip verification
+        if (result != null) return null;
+        
+        var checksumDownload = System.Text.Encoding.UTF8.GetString(checksumStr.ToArray());
+        var checksumCompute = Convert.ToHexString(await Hasher.ComputeHashAsync(destination));
+        
+        // Verification successful
+        if (checksumCompute == checksumDownload) return null;
+        // Verification failed, remove corrupted files
+        var parentDirectory = Path.GetFileName(Path.GetDirectoryName(url)) ?? "";
+        File.Delete(Path.Combine(Config.CacheDir, "Cache", parentDirectory, Path.GetFileName(url)));
+        File.Delete(Path.Combine(Config.CacheDir, "Cache", parentDirectory, Path.GetFileName(checksumUrl)));
+        return "Fetched corrupted file (checksum mismatch)";
+    }
```

Comment:
This assumes that the downloaded file is a zip and that a hash is present. The purpose of this function is to download any file the caller may request. The checksum seems unnecessary, as the only possibility of a "corruption" would be a returned error message (which should be indicated by the response status code).